Added new roms and comments to M20 hash file

eberhab opened the pull request:
Added more game roms to the M20 hash file. This should now be a more or less complete list of all the games that are available for the M20.

Also added a comment about a common issue with some of the already existing bootable system roms and how to fix it.

DavidHaywood commented:
are the disk images in question bad 'missing track 0' or actually just not bootable disks, that instead require you to boot from something else first?

if the former, any idea how that happened? it seems a remarkably specific problem, rather than one I'd expect to see repeated multiple times.

eberhab commented:
Yes, it is indeed a specific problem with this system, which uses mixture of FM/MFM encoding on the floppies. Usually PC controllers could not read the FM encoded first track, resulting in the 4kiB to be skipped during imaging. This is actually the case for the majority of the images. Floppies with a missing first track do not boot or boot only partially. More details can be found in the cited article in the comment. I have tried to write this down in even more detail here: https://github.com/eberhab/m20/blob/master/emulation_with_mame.md#the-anatomy-of-m20-floppies-and-mame-images

The new roms added with this PR do not have this problem. They all have an intact first track which was created using MAME itself.

Apologies for the second commit. The xml seemingly worked in Mame. I did not realize that there was a validate function. A few issues have been fixed. The file should now be in working order.

cracyc commented:
A little more discussion of this can be found at https://forums.bannister.org/ubbthreads.php?ubb=showflat&Number=100146#Post100146 .

eberhab commented:
Thanks for the questions and comments. I have added the link to the forum discussion to the comment about the track0 issue, since it is the original source. Anything else I can do to improve this?

Thanks for considering the addition. From my end this would be good to go.

cuavas commented on hash/m20.xml:
```diff
@@ -14,6 +14,11 @@ license:CC0-1.0
   to properly access volume 1 ("VL 1: " gives error)
   OTOH, I have no idea how to boot the disks with sits on PCOS prompt...
 
+  EDIT: Added more floppies with games. Most of the original floppies in this file
+  have a missing first track. This is one reason for the error messages observed in the note above.
+  Track 0 does not contain any user data and can be transferred from a working floppy (e.g. pcos102).
+  More details: https://forums.bannister.org/ubbthreads.php?ubb=showflat&Number=100146#Post100146
```
Don’t put revision history information here.  Please comment the individual disks with missing tracks and mark them as bad dumps.  Links to forum threads aren’t such a good idea.  If the missing tracks are due to FM not being readable in PC drives, say so.

eberhab replied:
Will do, I'll need some time to go through the existing images to comment them.

eberhab replied:
This should be solved.

cuavas commented on hash/m20.xml:
```diff
@@ -456,8 +461,8 @@ license:CC0-1.0
 		</part>
 	</software>
 
-
 <!-- This is a user disk with high school math programs -->
+
 	<software name="m20utent" supported="no"> <!-- Collection of BASIC programs, boot PCOS and run basic to use -->
```
Please don’t separate the comment from the item it belongs to like this.

eberhab replied:
done

cuavas commented on hash/m20.xml:
```diff
@@ -470,4 +475,630 @@ license:CC0-1.0
+        </part>
+    </software>
+
+    <software name="marilyn128k">
+        <description>Marilyn Monroe 128k</description>
+        <year>198?</year>
+        <publisher>Olivetti?</publisher>
+
+        <part name="flop1" interface="floppy_5_25">
+            <dataarea name="flop" size="286720">
+                <rom name="marilyn128k.img" size="286720" crc="ffd839ad" sha1="db6ff344c6f118992dd109b6700f7da74db14af4"/>
+            </dataarea>
+        </part>
+    </software>
+
+    <software name="marilyn">
+        <description>Marilyn Monroe</description>
+        <year>198?</year>
+        <publisher>Olivetti?</publisher>
+
+        <part name="flop1" interface="floppy_5_25">
+            <dataarea name="flop" size="286720">
+                <rom name="marilyn.img" size="286720" crc="b072cb66" sha1="841db31d0451a86f623915126f5b04fdf23c82df"/>
+            </dataarea>
+        </part>
+    </software>
```
Are these different versions of the same game?  If so, one should be a clone of the other.

eberhab replied:
Those are actually two completely different drawing demos, possibly highlighting different programming approaches to draw graphics on the M20.

eberhab replied:
I have removed this new image creation for now, since it is not reproducible/ matching the softlist philosophy.

cuavas commented on hash/m20.xml:
```diff
@@ -470,4 +475,630 @@ license:CC0-1.0
+        </part>
+    </software>
+
+    <software name="mauerschiessen">
+        <description>Mauerschießen</description>
```
Please keep short names terse.  Try to abbreviate them to eight characters if possible.

cuavas commented on hash/m20.xml:
```diff
@@ -470,4 +475,630 @@ license:CC0-1.0
+        </part>
+    </software>
+
+    <software name="othello_en">
+        <description>Othello (EN)</description>
+        <year>198?</year>
+        <publisher>Olivetti?</publisher>
+
+        <part name="flop1" interface="floppy_5_25">
+            <dataarea name="flop" size="286720">
+                <rom name="othello_en.img" size="286720" crc="b8b6f578" sha1="8e8d996addd1245ca0fe853e5ee7e49e4ec1e8ad"/>
+            </dataarea>
+        </part>
+    </software>
+
+    <software name="othello">
+        <description>Othello (D)</description>
+        <year>198?</year>
+        <publisher>Olivetti?</publisher>
+
+        <part name="flop1" interface="floppy_5_25">
+            <dataarea name="flop" size="286720">
+                <rom name="othello.img" size="286720" crc="79c904c0" sha1="66b4267c0c79b194cf81272a538354cd2127b63b"/>
+            </dataarea>
+        </part>
+    </software>
```
Please use parent/clone relationships, make the English version the parent, prefer to make clone short name by appending a suffix to the parent short name, and don’t unnecessarily use abbreviations in descriptions.

eberhab replied:
Actually the German version is the original and the English version is a more recent translation of the game. This brings up a more general question if these modern image creations are even suitable to be integrated into the software list? Would it be acceptable if those were marked as clones and additionally commented as e.g. a custom translation, or should those rather not be in the list since they are not "originals"? Given that we already have custom user disks in here, I find it difficult to gauge what is accepatble.

eberhab replied:
Added relationships and updated descriptions

cuavas commented on hash/m20.xml:
```diff
@@ -470,4 +475,630 @@ license:CC0-1.0
+        </part>
+    </software>
+
+    <software name="twodays1">
+        <description>Two Days to the Race (EN) - part 1</description>
+        <year>2019</year>
+        <publisher>D.Bucci</publisher>
+
+        <part name="flop1" interface="floppy_5_25">
+            <dataarea name="flop" size="286720">
+                <rom name="twodays1.img" size="286720" crc="f24ff7e0" sha1="4f6201ae7c6600fbbb44d5354d24817eb5b0ddfa"/>
+            </dataarea>
+        </part>
+    </software>
+
+    <software name="twodays2">
+        <description>Two Days to the Race (EN) - part 2</description>
+        <year>2019</year>
+        <publisher>D.Bucci</publisher>
+
+        <part name="flop1" interface="floppy_5_25">
+            <dataarea name="flop" size="286720">
+                <rom name="twodays2.img" size="286720" crc="d4662856" sha1="7a2b25408e69fb7acf59421d0f400d1bd00181bf"/>
+            </dataarea>
+        </part>
+    </software>
+
+    <software name="twodays3">
+        <description>Two Days to the Race (EN) - part 3</description>
+        <year>2019</year>
+        <publisher>D.Bucci</publisher>
+
+        <part name="flop1" interface="floppy_5_25">
+            <dataarea name="flop" size="286720">
+                <rom name="twodays3.img" size="286720" crc="1787191e" sha1="8755d37d3cc8a901098d33c3241d8af2a613f7f1"/>
+            </dataarea>
+        </part>
+    </software>
```
Are these actually separate products, or just different disks from one package?  In the latter case, they should be parts of the same item.

eberhab replied:
It's one package. What are the implications of the part-name field? The software list guideline says that it can be used as a command line argument. However all of these parts need to be loaded in "-flop1".

wilbertpol replied:
The -flop1 command line option and the part name are not linked they just both happen to contain 'flop1'. You can list them as '<part name="flop1"...', '<part name="flop2"...', etc. From the command line they can be loaded separately using something like '-flop1 twodays:flop3'

eberhab replied:
I have added the parent floppy, which is a single product, instead of the individual games, which were clones.

eberhab commented:
I have removed all the new game disc creations for now since they were not original floppies and I think in hindsight not suitable for the official sofware list. The main work of this PR is now:
*  Imaged a few more original floppies, including the FM track
* Checked existing images and indiviually mark them as good or bad
* Add some more individual comments about differing formats and where the disk is bootable

eberhab commented on hash/m20.xml:
```diff
@@ -337,23 +399,36 @@ license:CC0-1.0
 
-	<software name="olitutor" supported="no">
-		<description>OliTutor (Ita)</description>
+	<software name="olitesta" cloneof="olitest"><!-- bootable (pcos20) -->
```
I did some more digging and discovered that this image is actually a duplicate/ identical copy of olitest(imd). The checksum of the unpadded data part is identical. Since both versions are floating around, what's the guideline for this? I am inclined to remove the duplicate and maybe just leave a comment with the original.

eberhab replied:
Remove this, since it is a 100% duplicate in a different format (img/ imd).

cuavas commented on hash/m20.xml:
```diff
@@ -31,126 +33,164 @@ license:CC0-1.0
+	</software>
+
+	<software name="pcos30e" supported="partial"><!-- Missing FM track0 -->
 		<description>Olivetti PCOS 3.0e</description>
-		<year>19??</year>
+		<year>1983</year>
 		<publisher>Olivetti</publisher>
 
 		<part name="flop1" interface="floppy_5_25">
 			<dataarea name="flop" size="286720">
-				<rom name="pcos30e.img" size="286720" crc="ee2f9763" sha1="7f7dddb7815d0a4615ed4eac8f4709fb67688b46"/>
+				<rom name="pcos30e.img" size="228963" crc="ee2f9763" sha1="7f7dddb7815d0a4615ed4eac8f4709fb67688b46" status="baddump"/>
 			</dataarea>
 		</part>
 	</software>
 
-	<software name="pcos41a">
+	<software name="pcos30ea" cloneof="pcos30e">
+		<description>Olivetti PCOS 3.0e (Copy 1)</description>
+		<year>1983</year>
+		<publisher>Olivetti</publisher>
+
+		<part name="flop1" interface="floppy_5_25">
+			<dataarea name="flop" size="286720">
+				<rom name="pcos30e.td0" size="286720" crc="c6de0eec" sha1="2e1eb3d152a92ed1262899208ef298179e1b61b3"/>
+			</dataarea>
+		</part>
+	</software>
+
+	<software name="pcos30eb" cloneof="pcos30e">
+		<description>Olivetti PCOS 3.0e (Copy 2)</description>
+		<year>1983</year>
+		<publisher>Olivetti</publisher>
+
+		<part name="flop1" interface="floppy_5_25">
+			<dataarea name="flop" size="286720">
+				<rom name="pcos30e.imd" size="286720" crc="383a2131" sha1="62698a393328dc3ac1cdd5d72360623f741e54e1"/>
+			</dataarea>
+		</part>
+	</software>
```
Are “(Copy 1)” and “(Copy 2)” actually parts of the titles, or just for disambiguation?  If it’s just for disambiguation, use “(alt 1)” and “(alt 2)” for consistency with other software lists.

cuavas commented on hash/m20.xml:
```diff
@@ -31,126 +33,164 @@ license:CC0-1.0
 		<part name="flop1" interface="floppy_5_25">
 			<dataarea name="flop" size="286720">
-				<rom name="pcos_41a.img" size="286720" crc="f2fa9799" sha1="6d01e7ecc91211e588b4af3c77b972b20deee63b"/>
+				<rom name="pcos_41a.img" size="286720" crc="f2fa9799" sha1="6d01e7ecc91211e588b4af3c77b972b20deee63b" status="baddump"/>
 			</dataarea>
 		</part>
 		<part name="flop2" interface="floppy_5_25">
 			<feature name="part_id" value="Tools" />
 			<dataarea name="flop" size="286720">
-				<rom name="pcos_41a_tools.img" size="286720" crc="24936905" sha1="d27a2cae62fd62f07dce0e3df397d2dea4c1253b"/>
+				<rom name="pcos_41a_tools.img" size="286720" crc="24936905" sha1="d27a2cae62fd62f07dce0e3df397d2dea4c1253b" status="baddump"/>
 			</dataarea>
 		</part>
 	</software>
 
-	<software name="cpm8k">
+	<software name="cpm8k" supported="no"><!-- Missing FM track0, boot with -ramsize 512k -->
 		<description>CP/M 8000</description>
 		<year>19??</year>
 		<publisher>&lt;unknown&gt;</publisher>
+		<info name="usage" value="Launch with: -ramsize 512k" />
 
 		<part name="flop1" interface="floppy_5_25">
 			<dataarea name="flop" size="286720">
-				<rom name="cpm8k.img" size="286720" crc="e0f8dfe3" sha1="9d6a17af5ef2775ad985bface454926ae3f2a322"/>
+				<rom name="cpm8k.img" size="286720" crc="e0f8dfe3" sha1="9d6a17af5ef2775ad985bface454926ae3f2a322" status="baddump"/>
 			</dataarea>
 		</part>
 	</software>
 
-	<software name="msdos20" supported="no">
+	<software name="msdos20">
 	<!-- "This is a standard 360k MS-DOS disk image, 512 bytes/sector, 9 sectors, 2 sides, 40 tracks." -->
 	<!-- Use with BIOS 2.0 and the 8086 APB -->
-		<description>MS-DOS 2.0</description>
-		<year>19??</year>
+		<description>Olivetti MS-DOS 2.0</description>
+		<year>1983</year>
 		<publisher>Microsoft</publisher>
+		<info name="usage" value="Launch with: -bios 1" />
```
Please use the mnemonic BIOS name (`m20-20d` in this case I believe).  The numbers are too unstable.

cuavas commented on hash/m20.xml:
```diff
@@ -159,137 +199,150 @@ license:CC0-1.0
+			<dataarea name="flop" size="368640">
 				<rom name="adm5.imd" size="33619" crc="f7abfeee" sha1="ee432dc896f9b034c1f7ff3fac67cf5b8e003ff5"/>
 			</dataarea>
 		</part>
-	</software>
-
-	<software name="adm51" supported="no">
-		<description>ADM51 Terminal Emulator for MS-DOS</description>
-		<year>19??</year>
-		<publisher>&lt;unknown&gt;</publisher>
-
-		<part name="flop1" interface="floppy_5_25">
-			<dataarea name="flop" size="8192">
-				<rom name="adm51.imd" size="8192" crc="8916ea22" sha1="99fec256ae75c6569dff5bb8005f0e7bc301f4d8"/>
+		<part name="flop2" interface="floppy_5_25"><!-- Missing MFM track0 -->
+			<dataarea name="flop" size="368640">
+				<rom name="adm51.imd" size="8192" crc="8916ea22" sha1="99fec256ae75c6569dff5bb8005f0e7bc301f4d8" status="baddump"/>
 			</dataarea>
 		</part>
 	</software>
 
-<!-- Programs -->
-
-	<software name="asm" supported="no"> <!-- Invalid in M20 -->
+	<software name="asm" supported="partial"><!-- Missing FM track0 -->
 		<description>Olivetti Z8000 Assembler 2.1</description>
 		<year>19??</year>
 		<publisher>Olivetti</publisher>
 
 		<part name="flop1" interface="floppy_5_25">
 			<dataarea name="flop" size="286720">
-				<rom name="asm_r2.1.a.img" size="286720" crc="90b0fb9e" sha1="b250ff32d40293da446ce96b513b0000fb7814d1"/>
+				<rom name="asm_r2.1.a.img" size="286720" crc="90b0fb9e" sha1="b250ff32d40293da446ce96b513b0000fb7814d1" status="baddump"/>
 			</dataarea>
 		</part>
 	</software>
 
-	<software name="dietasm">
+	<software name="dietasm" supported="no"><!-- Missing FM track0, boot with -ramsize 512k -->
```
For software that requires more RAM than the default configuration, it’s better to add that in a `usage` `info` element so it’s visible in the UI, etc.

cuavas commented on hash/m20.xml:
```diff
@@ -159,137 +199,150 @@ license:CC0-1.0
 		</part>
 	</software>
 
 	<software name="dbms" supported="partial">
+		<!-- launch with > ba "dbms.bas" -->
 		<description>Olivetti Data Base Management System</description>
```
It’s better to put instructions for launching in `usage` `info` elements so it’s visible in the UI.

cuavas reviewed (commented):
Why have you changed the sizes of data areas?  It’s ignored by the “file” loaded used for floppy formats, but required to be at least as large as the size of the media image file for validation.  The convention is to set it to the actual size of the file.

eberhab commented on hash/m20.xml:
```diff
@@ -301,173 +354,275 @@ license:CC0-1.0
 	</software>
 
-	<software name="olisort" supported="no"> <!-- Collection of BASIC programs, boot PCOS and run basic to use -->
+	<software name="olisort" supported="no"><!-- Missing FM track0 -->
+		<!-- Collection of BASIC programs, boot PCOS and run basic to use -->
 		<description>OliSort 2.0.6</description>
 		<year>19??</year>
 		<publisher>Olivetti</publisher>
 
 		<part name="flop1" interface="floppy_5_25">
 			<dataarea name="flop" size="286720">
-				<rom name="olisort_206.img" size="286720" crc="08b7a3fd" sha1="704b1dd6211523835b6cf03ec6323e633315ce35"/>
+				<rom name="olisort_206.img" size="286720" crc="08b7a3fd" sha1="704b1dd6211523835b6cf03ec6323e633315ce35" status="baddump"/>
 			</dataarea>
 		</part>
 	</software>
 
-	<software name="olistat" supported="no"> <!-- Collection of BASIC programs, boot PCOS and run basic to use -->
+	<software name="olistat" supported="no"><!-- Missing FM track0 -->
+	<!-- Collection of BASIC programs, boot PCOS and run basic to use -->
 		<description>OliStat</description>
-		<year>19??</year>
+		<year>1981</year>
 		<publisher>Olivetti</publisher>
 
 		<part name="flop1" interface="floppy_5_25">
 			<dataarea name="flop" size="286720">
-				<rom name="olistat.img" size="286720" crc="10a68239" sha1="32427fd58cc3b07d6519d8dbc009dbcb450084e2"/>
+				<rom name="olistat.img" size="286720" crc="10a68239" sha1="32427fd58cc3b07d6519d8dbc009dbcb450084e2" status="baddump"/>
 			</dataarea>
 		</part>
 	</software>
 
-	<software name="olitest">
-		<description>OliTest</description>
-		<year>19??</year>
+	<software name="olitest"><!-- bootable pcos20 -->
+		<!-- olitest_7519913c.img is an identical img version of this disk --> 
+		<description>OliTest (Copy 1)</description>
+		<year>1981</year>
```
The olitest.img with crc 7519913c does exist. What's the recommendation here? Should be added or ignored?

eberhab commented on hash/m20.xml:
```diff
@@ -31,126 +33,164 @@ license:CC0-1.0
+
+		<part name="flop1" interface="floppy_5_25">
+			<dataarea name="flop" size="286720">
+				<rom name="pcos20h.img" size="286720" crc="55f470a7" sha1="7ac52fa8b52954410bc4ac496d6038a9a8204e8f" status="baddump"/>
+			</dataarea>
+		</part>
+	</software>
+
+	<software name="pcos30e" supported="partial"><!-- Missing FM track0 -->
 		<description>Olivetti PCOS 3.0e</description>
-		<year>19??</year>
+		<year>1983</year>
 		<publisher>Olivetti</publisher>
 
 		<part name="flop1" interface="floppy_5_25">
 			<dataarea name="flop" size="286720">
-				<rom name="pcos30e.img" size="286720" crc="ee2f9763" sha1="7f7dddb7815d0a4615ed4eac8f4709fb67688b46"/>
+				<rom name="pcos30e.img" size="228963" crc="ee2f9763" sha1="7f7dddb7815d0a4615ed4eac8f4709fb67688b46" status="baddump"/>
```
That one seems to be a typo, I will re-check the sizes.

eberhab replied:
Fixed

eberhab commented:
Concerning the file/ are sizes, are these conventions written down somewhere? The only guideline I could find does not contain info about the sizes: https://docs.mamedev.org/contributing/softlist.html 
I had changed the dataarea sizes to match the actual raw-sectors size. Will change the dataarea size to match the file sizes.

eberhab commented:
Again thanks for the thorough review. Updates:
- Converted comments into `usage` `info` tags
- Improved `alt` notation for consistency
- Changed `dataarea` `size` to match `rom` `size`
- Use bios version by name instead of number

Open questions:
- How to deal with (checksum-)identical images in different formats (e.g. imd vs. imd). Add both or only keep the (suspected) original?

cuavas commented:
> How to deal with (checksum-)identical images in different formats (e.g. imd vs. imd). Add both or only keep the (suspected) original?

I think in general it’s fine to keep just the best-quality dump when there are two dumps of identical disks.

> What qualifies a "working" image? Should we list images based on their current state, or assuming that the user has fixed the track0 issue?

It should be based on the current state.

Tafoid commented:
@eberhab 
Seeing as this has been accepted, the images should be made available to the dev team (or probably should have earlier).
In any case, please provide the data (or links to the data) so that this software list submission can be verified - using the email detailed [HERE](https://wiki.mamedev.org/index.php/Submitting_Source_Code).

Thanks for your contributions!
